src/vmware/vmxnet3_net.c

```diff
@@ -0,0 +1,581 @@
+
+    vmxnet3 vn = allocate(dev->general, sizeof(struct vmxnet3));
+    vn->dev = dev;
+    vn->n = allocate(dev->general, sizeof(struct netif));
+
+    vn->rxbuflen = VMXNET3_RX_MAXSEGSIZE;
+    vn->rxbuffers = allocate_objcache(dev->general, page_allocator,
+                      vn->rxbuflen + sizeof(struct xpbuf), PAGESIZE_2M);
+
+    dev->vmx_ds = allocate_zero(dev->contiguous, sizeof(struct vmxnet3_driver_shared));
+
+    vn->rx_intr_handler = closure(dev->general, rx_interrupt, vn);
+    pci_setup_msix(dev->dev, 0, vn->tx_intr_handler, "vmxnet3 tx");
+    pci_setup_msix(dev->dev, 1, vn->rx_intr_handler, "vmxnet3 rx");
+
+    vmxnet3_tx_queues_alloc(dev);
+    vmxnet3_rx_queues_alloc(dev);
+
+    vmxnet3_queues_shared_alloc(dev);
+
+    for (int i = 0; i < VMXNET3_RXRINGS_PERQ; i++) {
+        for (int idx = 0; idx < VMXNET3_MAX_RX_NDESC; idx++)
+            vmxnet3_newbuf(vn, i);
+    }
+
+    vmxnet3_init_shared_data(dev);
+
+    /* Check device versions */
+    vmxnet3_check_version(dev);
+
+    vmxnet3_set_interrupt_idx(dev);
+
+    vmxnet3_interrupts_disable(dev);
+
+    /* enable device */
+    init_vmxnet3_driver_shared(dev);
+    test_shared(vn);
+    vmxnet3_read_cmd(dev, VMXNET3_CMD_ENABLE);
+    pci_bar_write_4(&dev->bar0, VMXNET3_BAR0_RXH1(0), 0);
+    pci_bar_write_4(&dev->bar0, VMXNET3_BAR0_RXH2(0), 0);
+
+    netif_add(vn->n,
+              0, 0, 0,
+              vn,
+              vmxif_init,
+              ethernet_input);
+
+    vmxnet3_interrupts_enable(dev);
+}
+
+closure_function(2, 1, boolean, vmxnet3_net_probe,
+                 heap, general, heap, page_allocator,
+                 pci_dev, d)
+{
+    if (!vmxnet3_probe(d))
+        return false;
+
+    vmxnet3_net_debug("FOUND VMXNET3!\n");
+    vmxnet3_net_attach(bound(general), bound(page_allocator), d);
+    vmxnet3_net_debug("ATTACHED VMXNET3!\n");
+    return true;
+}
+
+void
+init_vmxnet3_network(kernel_heaps kh)
+{
+    heap h = heap_general(kh);
+    register_pci_driver(closure(h, vmxnet3_net_probe, h, heap_backed(kh)));
+}
+
+static void
+vmxnet3_discard(vmxnet3_pci dev, int rid, int idx)
+{
+    struct vmxnet3_rxqueue *rxq = dev->vmx_rxq[0];
+    struct vmxnet3_rxring *rxr = &rxq->vxrxq_cmd_ring[rid];
+    struct vmxnet3_rxdesc *rxd = &rxr->vxrxr_rxd[idx];
+    rxd->gen = rxr->vxrxr_gen;
+    if (++rxr->vxrxr_refill_start == rxr->vxrxr_ndesc) {
+        rxr->vxrxr_refill_start = 0;
+        rxr->vxrxr_gen ^= 1;
+    }
+}
+
+static void
+receive_buffer_release(struct pbuf *p)
+{
+    xpbuf x  = (void *)p;
+    deallocate(x->vn->rxbuffers, x, x->vn->rxbuflen + sizeof(struct xpbuf));
+}
+
+void
+vmxnet3_newbuf(vmxnet3 vdev, int rid)
+{
+    vmxnet3_pci dev = vdev->dev;
+    struct vmxnet3_rxqueue *rxq = dev->vmx_rxq[0];
+    struct vmxnet3_rxring *rxr = &rxq->vxrxq_cmd_ring[rid];
+
+    int idx = rxr->vxrxr_refill_start;
+    struct vmxnet3_rxdesc *rxd = &rxr->vxrxr_rxd[idx];
+
+    xpbuf x = allocate(vdev->rxbuffers, sizeof(struct xpbuf) + vdev->rxbuflen);
```

Comment:
Allocating here may be racy if this function is called at interrupt level. lwIP processing occurs in bottom half processing, so a dealloc to vdev->rxbuffers there may clash with an allocate here.

In general, allocations from system memory should never happen at interrupt level. But since this is a private cache of pbufs, a spin_lock_irq around the dealloc and spin_lock here may suffice.